src/pybind/mgr/rbd_support/mirror_snapshot_schedule.py

```diff
@@ -0,0 +1,680 @@
+
+    def create_snapshot(self, pool_id, namespace, image_id):
+        try:
+            with self.module.rados.open_ioctx2(int(pool_id)) as ioctx:
+                ioctx.set_namespace(namespace)
+                with rbd.Image(ioctx, image_id=image_id) as image:
+                    mode = image.mirror_image_get_mode()
+                    if mode != rbd.RBD_MIRROR_IMAGE_MODE_SNAPSHOT:
+                        return
+                    info = image.mirror_image_get_info()
+                    if info['state'] != rbd.RBD_MIRROR_IMAGE_ENABLED or \
+                       not info['primary']:
+                        return
+                    snap_id = image.mirror_image_create_snapshot()
+                    self.log.debug(
+                        "create_snapshot: {}/{}/{}: snap_id={}".format(
+                            ioctx.get_pool_name(), namespace, image.get_name(),
+                            snap_id))
+        except Exception as e:
+            self.log.error(
+                "exception when creating snapshot for {}/{}/{}: {}".format(
+                    pool_id, namespace, image_id, e))
+
+
+    @classmethod
+    def format_image_spec(cls, image_spec):
+        image = image_spec[2]
+        if image_spec[1]:
+            image = "{}/{}".format(image_spec[1], image)
+        if image_spec[0]:
+            image = "{}/{}".format(image_spec[0], image)
+        return image
+
+    def init_schedule_queue(self):
+        self.queue = {}
+        self.images = {}
+        self.refresh_images()
+        self.log.debug("scheduler queue is initialized")
+
+    def load_schedules(self):
+        self.log.info("MirrorSnapshotScheduleHandler: load_schedules")
+
+        schedules = {}
+        schedule_cfg = self.module.get_localized_module_option(
+            'mirror_snapshot_schedule', '')
+        if schedule_cfg:
+            try:
+                schedule = Schedule.from_json('/', schedule_cfg)
+                schedules['/'] = schedule
+            except ValueError:
+                self.log.error("Failed to decode configured schedule {}".format(
+                    schedule_cfg))
+
+        for pool_id, pool_name in get_rbd_pools(self.module).items():
+            with self.module.rados.open_ioctx2(int(pool_id)) as ioctx:
+                self.load_pool_schedules(ioctx, schedules)
+
+        with self.lock:
+            self.schedules = schedules
+
+    def load_pool_schedules(self, ioctx, schedules):
+        pool_id = ioctx.get_pool_id()
+        pool_name = ioctx.get_pool_name()
+        stale_keys = []
+        start_after = ''
+        try:
+            while True:
+                with rados.ReadOpCtx() as read_op:
+                    self.log.info("load_schedules: {}, start_after={}".format(
+                        pool_name, start_after))
+                    it, ret = ioctx.get_omap_vals(read_op, start_after, "", 128)
+                    ioctx.operate_read_op(read_op, SCHEDULE_OID)
```

Comment:
I am not sure I like the idea to store `SCHEDULE_OID` in the associated namespace. It is not very helpful for the case you described (when a user does not have access to the global namespace), because it needs it anyway, for the case when the scheduler is not specified on the image or namespace level, to get the scheduler on a higher level.

Currently cleaning the stale schedules after the namespace removal does not look like a big deal, and we still have to use this approach to remove stale schedules on image level. I think the less objects we need to read to build the schedule table the better. Because otherwise why not store image level schedules in image metadata?

Anyway, I am fine with any option, so if you still prefer `SCHEDULE_OID` per namespace I will update it.

Reply:
I guess that is true -- this MGR module will be using its own credentials so it doesn't matter that the user doesn't have access to the global namespace. We probably should add hooks for `is_authorized` to ensure the user has the proper caps.